### PR TITLE
Bugfix FXIOS-7127 Preserve URL on tab close and restore for web apps

### PR DESCRIPTION
**Pull request by ih-codes**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7127)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/15837)

## :bulb: Description
When a tab is closed and the Undo button is pressed to restore it from the bottom toast, there's an incorrect redirect of the URL for web apps using pushState (e.g. YouTube). 

For example, if you visit YouTube, tap the top video, and then close the tab and undo closing the tab, the webView should remain on the selected video. However, prior to this bug fix, the restored tab would be back on the home page of YouTube.

Note that closing and restoring the last tab has an incorrect selection state in the tab tray. That’s a separate bug here https://github.com/mozilla-mobile/firefox-ios/issues/20850 (PR https://github.com/mozilla-mobile/firefox-ios/pull/20854)
 
### Root Cause

This is an asynchronicity issue. The webView's session history was not being saved (background Task) in time before the webView was destroyed for the closed tab. So when the tab was restored, the [interactionState](https://developer.apple.com/documentation/webkit/wkwebview/3752236-interactionstate) (or `tabSession` as we call it) was unavailable.

### Testing

This fix should be tested significantly since there are a lot of edge cases. (I didn't test tab syncing in depth, but I did make another bug ticket for a similar issue with syncing behaviour: [GitHub](https://github.com/mozilla-mobile/firefox-ios/issues/20904) [JIRA](https://mozilla-hub.atlassian.net/browse/FXIOS-9447))

**Test Paths:**

1) Tab Refactor Enabled
2) Tab Refactor Disabled

**Test Cases for Each Path:**

1) Close and Undo Close on a single tab
2) Close and Undo Close on the first tab, a middle tab, and a last tab when multiple tabs are open (2, 3+)
2.5) Do this for both the active tab and an inactive tab
3) Close and Undo Close on a tab using the tab button long press submenu
4) Close all tabs at once, and restore all tabs

Note: For all these test cases, the tab in question must be the active tab. Switching tabs and then closing that tab seemed to already save the state as expected. But it wouldn't hurt to test again.

### Testing Link

You can use this link for testing instead of YouTube: https://stuff.overengineer.dev/mozilla/testcases/pushstate-nav/0.html
(courtesy of https://github.com/mozilla-mobile/firefox-ios/issues/15837#issuecomment-2162552585)

## Fix Demo

https://github.com/mozilla-mobile/firefox-ios/assets/173110554/b9c938f3-f533-416b-9275-33ec9d9385e8

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by ih-codes** [comment]
@OrlaM @dataports Please take a look and see if anything goes against your tab tray refactor (I don't quite have the full picture of that yet). I have a `TODO` in here that might need some attention--it specifically fixed the case of closing a single tab when using the legacy tab manager.

Also, I wonder if this *might* be related to @mattreaganmozilla 's note here https://github.com/mozilla-mobile/firefox-ios/pull/20854#issuecomment-2207381450 since the root cause of the issue appeared to be a Task dispatch delaying the webview's interactionState save in `saveCurrentTabSessionData()`.

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 1]
```diff
@@ -626,7 +632,8 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
 
     func removeTabs(_ tabs: [Tab]) {
         for tab in tabs {
-            self.removeTab(tab, flushToDisk: false)
+            // TODO Is flushToDisk true safe here? Needed for tab refactor disabled (legacy), single tab removal saving state
+            self.removeTab(tab, flushToDisk: true)
```
Comment (dataports) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:636]: Hmm, this solves the problem but I think it introduces performance issues when we close many tabs with the way the code is functioning currently on the tab tray refactor. It is quite slow if I open up 50+ test tabs. I experimented with dismissing the tab tray earlier but this introduces some funny visual effects while it tries to save so many tabs session data, and with the way the code is now in main you would just wait a long time for everything to close in the tab tray before it's dismissed.

I'm going to think about how to solve this some more this week and see if I can think of a solution to discuss with Orla next week when she's back, seems like the synchronicity issue is kind of the bottleneck here, but I'm not sure.

For reference, this is what it looks like even if I dismiss the tab tray earlier.
https://github.com/mozilla-mobile/firefox-ios/assets/5545720/8013ac7b-78b3-41c9-8af3-7aab8aede571

Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:636]: Thanks for pointing that out @dataports ! When I test this case on main I don't see the flickering but the tab tray also hangs for a while without feedback.

Since this specific line only fixes one edge case (single tab closed and restored in the legacy implementation, it could be removed especially if the intention is to eventually roll out the entire tab tray refactor.

Let me know if you could use any help digging into the synchronicity issue. Like Matt's PR, my original solution was looking better before that one PR got rolled back so it seems to have something to do with the change to an asynchronous way of doing things.

One of my alternative ideas for this bug was to start adding completion handlers or awaiting asynchronous session saves, but I felt like I was fighting against whatever the current pattern was and looked for a less invasive solution.

Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:636]: @dataports Just was returning to this branch since the synchronous tab selection PR got re-merged.

I actually get the exact behaviour from your sample video on `main` right now, so it does not appear to be something caused by this PR's changes. 🤔  Should we make a performance related ticket for this?

On a separate note, I'll back out of this specific line since we are no longer supporting the legacy tab manager.

Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift]: Made a ticket for the flashing tab close [JIRA](https://mozilla-hub.atlassian.net/browse/FXIOS-9474) and [Github](https://github.com/mozilla-mobile/firefox-ios/issues/20967)

Comment (dataports) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift]: Hmm, probably another issue related to `favicons` 😭 generally we have bugs open for them flashing in a bunch of contexts

**Review by OrlaM — APPROVED** [review]
LGTM
